File: task22/task22.py
def find_smallest(nums):
    result = []
    s = []

    for x in nums:
        while len(s) != 0 and s[-1] >= x:
            s.pop()
        result.append(s[-1] if len(s) != 0 else -1)

        s.append(x)

    return result


def test1():
    t = [2, 5, 3, 7, 8, 1, 9]
    print(find_smallest(t) == [-1, 2, 2, 3, 7, -1, 1])


def test2():
    t = [5, 7, 4, 9, 8, 10]
    print(find_smallest(t) == [-1, 5, -1, 4, 4, 8])

def test3():
    t = [1, 5, 2, 2, 2, 5, 5, 4]
    print(find_smallest(t) == [-1, 1, 1, 1, 1, 2, 2, 2])

def main():
    test1()
    test2()
    test3()


if __name__ == '__main__':
    main()

File: archive/task22/task22.cpp
#include <iostream>

int *find_smallest(const int *input, const int size)
{
    int *result = new int[size];

    for (int i = 0; i < size; i++)
    {
        bool found = false;

        for (int j = i - 1; j >= 0; --j)
        {
            if (input[j] < input[i])
            {
                result[i] = input[j];
                found = true;
                break;
            }
        }

        if (!found)
        {
            result[i] = -1;
        }
    }

    return result;
}

void test1()
{
    int size = 7;
    int input[] = {2, 5, 3, 7, 8, 1, 9};
    int *result = find_smallest(input, size);

    for (int i = 0; i < size; i++)
    {
        std::cout << result[i] << ' ';
    }
    std::cout << '\n';

    delete[] result;
    result = nullptr;
}

void test2()
{
    int size = 6;
    int input[] = {5, 7, 4, 9, 8, 10};
    int *result = find_smallest(input, size);

    for (int i = 0; i < size; i++)
    {
        std::cout << result[i] << ' ';
    }
    std::cout << '\n';

    delete[] result;
    result = nullptr;
}

void test3()
{
    int size = 8;
    int input[] = {1, 5, 2, 2, 2, 5, 5, 4};
    int *result = find_smallest(input, size);

    for (int i = 0; i < size; i++)
    {
        std::cout << result[i] << ' ';
    }
    std::cout << '\n';

    delete[] result;
    result = nullptr;
}

int main(int argc, char const *argv[])
{
    test1();
    test2();
    test3();

    return 0;
}
